2274 cypress user approval

elipe17 opened the pull request:
## Summary of Changes
- Implemented tests for user approval flow
- Updated initial test structure to include common test steps
- Made updates to existing docs where necessary
Pull request closes #2274 

## How to Test
```
# in separate terminals
cd tdrs-frontend && npm run start
cd tdrs-backend && docker-compose up

# in another terminal
docker-compose exec web python manage.py delete_cypress_users -usernames new-cypress@teamraft.com cypress-admin@teamraft.com
docker-compose exec web python manage.py generate_cypress_users
cd ../tdrs-frontend
npm run test:e2e
```

## Deliverables

### [Deliverable 1: Accepted Features](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-1-Accepted-Features)

Checklist of ACs:
+ [x] Integration tests cover critical path of automated user being approved
+ [x] Test flow can be run both in CircleCI and deployed environments
+ [x] Testing Checklist has been run and all tests pass
+ [x] README updated
- [x] `lfrohlich` and/or `adpennington` confirmed that ACs are met.

### [Deliverable 2: Tested Code](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-2-Tested-Code)

+ Are all areas of code introduced in this PR meaningfully tested?
  + [x] If this PR introduces backend code changes, are they meaningfully tested?
  + [x] If this PR introduces frontend code changes, are they meaningfully tested?
+ Are code coverage minimums met?
  + [x] Frontend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)
  + [x] Backend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)

### [Deliverable 3: Properly Styled Code](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-3-Properly-Styled-Code)

+ [x] Are backend code style checks passing on CircleCI?
+ [x] Are frontend code style checks passing on CircleCI?
+ [x] Are code maintainability principles being followed?

### [Deliverable 4: Accessible](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-4-Accessibility)

+ [ ] Does this PR complete the epic? 
+ [ ] Are links included to any other gov-approved PRs associated with epic?
+ [ ] Does PR include documentation for Raft's a11y review? 
+ [ ] Did automated and manual testing with `iamjolly` and `ttran-hub` using Accessibility Insights reveal any errors introduced in this PR?


### [Deliverable 5: Deployed](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-5-Deployed)

+ [ ] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?

### [Deliverable 6: Documented](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-6-Code-documentation)

+ [x] Does this PR provide background for why coding decisions were made?
+ [x] If this PR introduces backend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [x] If this PR introduces frontend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [x] If this PR introduces dependencies, are their licenses documented?
+ [x] Can reviewer explain and take ownership of these elements presented in this code review?

### [Deliverable 7: Secure](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-7-Secure)

+ [x] Does the OWASP Scan pass on CircleCI?
+ [ ] Do manual code review and manual testing detect any new security issues?
+ [ ] If new issues detected, is investigation and/or remediation plan documented? 

### [Deliverable 8: User Research](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-8-User-Research)

Research product(s) clearly articulate(s):
+ [ ] the purpose of the research
+ [ ] methods used to conduct the research 
+ [ ] who participated in the research
+ [ ] what was tested and how
+ [ ] impact of research on TDP
+ [ ] (_if applicable_) final design mockups produced for TDP development

Commit:
- initial commit for admin approving user

Commit:
- Updated to support backend admin api requests
- Updating tests to approve a new user

Commit:
- Updated steps to get cypress user approved
- Added 'Given' step to ensure the cypress user is always in the correct state at the start of a test

Commit:
- Added request state
- adding missing fields to admin post request

elipe17 committed:
- adding logic for user to request access

elipe17 committed:
- Added env variables to minimize hardcoded config
- Added logic to approve user in request state

jtimpe committed:
fix cypress csrf errors

elipe17 committed:
- Cleaned up code
- Renamed common steps

elipe17 committed:
- Updated circi job to create super user
- updated docs
- quick rename of common steps
- updated glob patterns for step file discovery

elipe17 committed:
- Removing debug statement

elipe17 committed:
- Fixed linting errors

elipe17 committed:
- removing unnecessary logging message

elipe17 committed:
- Added tests for denying and deactivating users

elipe17 committed:
- increased the wait for the page lload to avoid accidental fails
- uncommented tests

elipe17 committed:
- Added final test for last acount status state

elipe17 committed:
- Renamed steps to keep consistency
- Updated indentation level to be 4

jtimpe commented on tdrs-backend/tdpservice/users/api/login.py:
```diff
@@ -416,4 +416,9 @@ def post(self, request):
 
-        return Response({'authenticated': True})
+        response = {'authenticated': True}
+        if "admin" in username:
```
can we do `if u.is_superuser:` here?

jtimpe commented on tdrs-backend/tdpservice/users/management/commands/generate_cypress_superuser.py:
```diff
@@ -0,0 +1,33 @@
+    def handle(self, *args, **options):
+        """Generate test users if they don't exist."""
+        if settings.DEBUG:
+            get_or_create_superuser('cypress-admin@teamraft.com')
```
can we add this as a line to the other `generate_cypress_users` file? that way there's only one command to add all the users that are needed. you can move the whole `get_or_create_super` function to the other file too!

jtimpe commented on tdrs-frontend/cypress/e2e/accounts/accounts.js:
```diff
@@ -0,0 +1,79 @@
+})
+
+When('The user requests access', () => {
+  cy.get('#firstName').type(Cypress.env('cypressName'))
```
i don't like using `Cypress.env` variables here, we should reserve those for configuring the test suite. i think just hardcoding the value is sufficient and more clear

jtimpe commented on tdrs-frontend/cypress/support/commands.js:
```diff
@@ -26,26 +26,115 @@
+  }
+})
+
+Cypress.Commands.add('reinitUserAccount', (user, token) => {
```
i like the abstraction of `reinitUserAccount`, but i don't think it needs to be a cypress command override. could we move the `options` declaration to `Cypress.Commands.add('adminApiRequest')` with parameters to set the `method`, `url`, and `body`? Then the `cy.adminApiRequest` call should be small enough to fit inside the step implementations, or we could make a parameterized common step - happy to work with you on that. a simple function separated from the cypress commands would also work well.

elipe17 replied:
I would like to discuss this with you a little further.

jtimpe commented on tdrs-frontend/package.json:
```diff
@@ -105,7 +105,8 @@
   "cypress-cucumber-preprocessor": {
     "stepDefinitions": [
-      "cypress/**/*.js"
+      "cypress/e2e/*/[filepath].js", 
```
i like this change. if i'm reading correctly, this will force each test's step implementations to be named the same as the feature file (i.e., `accounts.feature` will pull from `accounts.js`), and will create a `common-steps` folder we can add shared steps to.

jtimpe commented on tdrs-frontend/src/actions/requestAccess.js:
```diff
@@ -14,9 +15,7 @@ export const requestAccess =
-      const { data } = await (
-        await axiosInstance
-      ).patch(URL, user, {
+      const { data } = await axios.patch(URL, user, {
```
just wanted to point out why this change was needed. `axiosInstance`, for whatever reason, didn't preserve the default `axios` behavior of injecting the `csrftoken` cookie into the `X-CSRFTOKEN` header. swapping it out here fixes the issue. I don't know where else `axiosInstance` is used but it will likely face the same issue with our cypress tests. Luckily, there doesn't seem to be any configuration on it, so swapping it out for the default should be safe (and seemingly adds functionality 😄 )

jtimpe commented on tdrs-frontend/cypress/e2e/common-steps/common-steps.js:
```diff
@@ -0,0 +1,18 @@
+  cy.adminLogin('cypress-admin@teamraft.com')
+})
+
+When('The user logs in', () => {
```
these are great common steps. if you wanted to parameterize the admin requests to update the account status, you could add them here with something like

`When('the admin sets approval status of {string} to {string}', (username, status) => {})`

used like this
`When the admin sets the approval status of "new-cypress@teamraft.com" to "Approved"`

rough example but hopefully that gives you an idea. https://github.com/badeball/cypress-cucumber-preprocessor/blob/master/docs/cucumber-basics.md

jtimpe commented on tdrs-frontend/cypress/e2e/accounts/accounts.feature:
```diff
@@ -1,5 +1,41 @@
+        Given The admin logs in
+        And The user is in begin state
+        When The user visits the home page
+        And The user logs in
```
these are pretty well written steps, you've made the tests more declarative which is great
https://cucumber.io/docs/bdd/better-gherkin/

i think we could improve here a little by making the "user" at play more obvious, either by naming them as in the docs examples above, or parameterizing some of the shared steps to pass the username in (i.e., `When "new-cypress" logs in`)

jtimpe commented on tdrs-frontend/cypress.config.js:
```diff
@@ -6,10 +6,15 @@ module.exports = defineConfig({
   e2e: {
     baseUrl: 'http://localhost:3000',
     specPattern: '**/*.feature',
+    experimentalSessionAndOrigin: true,
```
what's this do?

EricLipe17 replied:
This can be removed now. I needed it for some other feature I was testing. It turned out not to be necessary and I forgot to remove it.

elipe17 replied:
P.S., I have no clue how I made the above comment with my personal GH. I logged out and logged back in to make sure I am in the correct account.

jtimpe reviewed (changes requested):
Really great overall, and I want to commend you again for being persistent and a good sport as we threw you way in the deep end for your first ticket. I'm mainly requesting that you move a couple things around such as the `Cypress.Commands` for `reinitUserAccount` and `changeUserInfo` and the `generate_cypress_superuser` django command; and remove the test data values from `cypress.config.js`. Up to you whether you want to play with the step parameterization or not.

elipe17 committed:
- Made updates based on review feedback

andrew-jameson commented:
Running the provided testing steps appears to be different than originally intended. It doesn't look harmful but is this just an oversight?

```
 tdrs-backend % docker-compose exec web python manage.py generate_cypress_users
created new-cypress@teamraft.com
created super user cypress-admin@teamraft.com
 tdrs-backend % docker-compose exec web python manage.py generate_cypress_superuser
Unknown command: 'generate_cypress_superuser'. Did you mean generate_cypress_users?
Type 'manage.py help' for usage.
```

Otherwise, tests run correctly.

elipe17 commented:
> Running the provided testing steps appears to be different than originally intended. It doesn't look harmful but is this just an oversight?
> 
> ```
>  tdrs-backend % docker-compose exec web python manage.py generate_cypress_users
> created new-cypress@teamraft.com
> created super user cypress-admin@teamraft.com
>  tdrs-backend % docker-compose exec web python manage.py generate_cypress_superuser
> Unknown command: 'generate_cypress_superuser'. Did you mean generate_cypress_users?
> Type 'manage.py help' for usage.
> ```
> 
> Otherwise, tests run correctly.

Sorry, that is an oversight on my end. I forgot to update the execution instructions after making some of the changes Jan suggested. There is no need to run `docker-compose exec web python manage.py generate_cypress_superuser`. The original `docker-compose exec web python manage.py generate_cypress_users` now takes care of that. Will update now.

jtimpe committed:
refactor adminApiRequest

raftmsohani commented on tdrs-backend/tdpservice/users/management/commands/generate_cypress_users.py:
```diff
@@ -21,6 +21,20 @@ def get_or_create_user(username):
+
+    try:
+        super_user = User.objects.get(username=username)
+        print(f'found {username}')
```
should we log instead of print?

raftmsohani commented on tdrs-frontend/cypress.config.js:
```diff
@@ -9,6 +9,7 @@ module.exports = defineConfig({
 
     env: {
       apiUrl: 'http://localhost:8080/v1',
+      adminUrl: 'http://localhost:8080/admin',
```
Nit: suggest to use env vars instead of hardcoding the urls.

elipe17 replied:
These are env vars which can be overriden in `package.json`. Is that sufficient or do you think they should be accessible from something like `process.env.adminUrl`?

elipe17 committed:
- Removing unnecessary function
- using adminApiRequest throughout

elipe17 committed:
- Removing groups and region on reinit

elipe17 committed:
- re-enabling all tests

elipe17 committed:
- Update CI job to refresh cypress users on start
- Updated docs
- Added command to delete cypress users
- Updated feature file to consistently reference user by name instead of "the user"
- Made prints in Commands log statements

elipe17 committed:
- Fixed lint errors

elipe17 committed:
- small change to gherkin to make more readable

jtimpe committed:
Merge branch 'develop' into 2274-cypress-user-approval

ADPennington commented:
@elipe17 can you resolve the merge conflicts here when you get a chance?

elipe17 committed:
Merge branch 'develop' of https://github.com/raft-tech/TANF-app into 2274-cypress-user-approval

elipe17 committed:
Merge branch '2274-cypress-user-approval' of https://github.com/raft-tech/TANF-app into 2274-cypress-user-approval

elipe17 committed:
- removing e2e command

elipe17 committed:
- Testing suggested update for nginx to pass the csrf token to django

elipe17 committed:
- Adding referer policy

elipe17 committed:
Revert "- Adding referer policy"

This reverts commit 4b32007539e560799f0c0e6ce8e6ef75b8b80bcf.

elipe17 committed:
Revert "- Testing suggested update for nginx to pass the csrf token to django"

This reverts commit ca2349bddff52c7d5f6eb6c58a296984f27f0fc5.

elipe17 committed:
- Testing nginx config to pass csrf token

elipe17 committed:
- Updating urls for new nginx config

elipe17 committed:
Merge branch 'develop' of https://github.com/raft-tech/TANF-app into 2274-cypress-user-approval

ADPennington commented:
@elipe17 can you resolve this merge conflict when you get a chance?

elipe17 committed:
Merge branch 'develop' of https://github.com/raft-tech/TANF-app into 2274-cypress-user-approval

elipe17 committed:
Merge branch 'develop' of https://github.com/raft-tech/TANF-app into 2274-cypress-user-approval

elipe17 committed:
- Added longer wait for request submitted view
- Updated csrf header key
- Getting nginx to pass csrf header along with cookies

ADPennington committed:
Merge branch 'develop' into 2274-cypress-user-approval
